cpg_utils/cloudpath_hail_az.py

```diff
@@ -4,26 +4,87 @@
 """
 
 import re
-from typing import Union, Optional
+import os
+import json
+import logging
+from typing import Union, Optional, Any, Callable
 from urllib.parse import urlparse
-
+import mimetypes
+from azure.identity import DefaultAzureCredential, ClientSecretCredential
+from azure.storage.blob import BlobServiceClient
+#import azure.storage.blob
 from cloudpathlib import AzureBlobClient, AzureBlobPath
 from cloudpathlib.client import register_client_class
 from cloudpathlib.cloudpath import register_path_class, CloudPath
 from cloudpathlib.exceptions import InvalidPrefixError
 
+# TODO pretty heavy handed
+logging.getLogger('azure').setLevel(logging.WARN)
+logging.getLogger('msal').setLevel(logging.WARN)
+logging.getLogger('urllib3').setLevel(logging.WARN)
 
 @register_client_class('hail-az')
 class HailAzureBlobClient(AzureBlobClient):
-    """
-    Client is identical to the base Azure implementation.
-    """
+    def __init__(
+        self,
+        account_url: str,
+        local_cache_dir: Optional[Union[str, os.PathLike]] = None,
+        content_type_method: Optional[Callable] = mimetypes.guess_type,
+    ):
+        """Class constructor. Sets up a [`BlobServiceClient`](
+        https://docs.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.blobserviceclient?view=azure-python).
+        Supports the following authentication methods, specific to Hail Batch.
+        - Implicit instantiation of a [`DefaultAzureCredential`](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet-preview)
+        - Environment variable `"AZURE_APPLICATION_CREDENTIALS"` containing a path to a JSON file that in turn contains three fields: 
+        "tenant", "appId", and "password".
+        - Account URL via `account_url`, authenticated with an embedded SAS token. In all cases an Account URL is required, and if a SAS token is provided, it will be used for Authentication.
+        If multiple methods are used, priority order is reverse of list above (later in list takes
+        priority).
+        Args:
+            account_url (str): The URL to the blob storage account, optionally
+                authenticated with a SAS token. See documentation for [`BlobServiceClient`](
+                https://docs.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.blobserviceclient?view=azure-python).
+            local_cache_dir (Optional[Union[str, os.PathLike]]): Path to directory to use as cache
+                for downloaded files. If None, will use a temporary directory.
+            content_type_method (Optional[Callable]): Function to call to guess media type (mimetype) when
+                writing a file to the cloud. Defaults to `mimetypes.guess_type`. Must return a tuple (content type, content encoding).
+        """
+        if account_url is None:
+            raise ValueError("account_url must be specified")
+
+        parsed = urlparse(account_url)
```

Comment:
the result here seems to be a complicated named tuple, which may affect your truthiness check? might be easier to deal with it unpacked - i.e. _, _, _, _, query, _ = ...